ENH: interpolate: make NdBSpline and RGI Array API compatible

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

towards gh-23754

#### What does this implement/fix?
<!--Please explain your changes.-->

Make `NdBSpline` array API compatible. The pattern is the same as `BSpline`: convert inputs to numpy, call the compiled implementation, convert back.

One exception is `design_matrix` which returns a `csr_array`, which is numpy-only, so there's nothing we can do until there's array api compatibility in `scipy.sparse`.

#### Additional information
<!--Any additional information you think is important.-->

ev-br committed:
ENH: interpolate: make NdBSpline Array API compatible

ev-br committed:
TST: interpolate/RGI: array api testing for RGI

ev-br committed:
ENH: interpolate: make RGI array api compatible

ev-br committed:
ENH: use xp-capabilities for RGI and NdBSpline

steppi commented on scipy/interpolate/_rgi.py:
```diff
@@ -281,17 +290,25 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+            xp = array_namespace(*points, values)
+        except Exception:
+            # "duck-type" values
+            xp = np_compat
```
I'm a little concerned about catching the bare `Exception` here and falling back to `np_compat`. The intention is that `values` only needs to quack like a duck, but if someone accidentally mixes array types in `points`, they may be surprised to find everything seems to work, but the NumPy backend is being used.

Might it be fine to instead do

```python
xp = array_namespace(*points)
```
?

All tests pass locally for me still with this change.

lucascolley replied:
I would err towards just `array_namespace(*points, values)` and make users adapt, but I agree that this suggestion is an improvement over silently using NumPy.

lucascolley replied:
> users should be given a deprecation period to adapt,

This is all hidden behind the experimental env var still, no? Without it, we get `np_compat` back from `array_namespace` anyway.

steppi replied:
Here's an example of one

https://github.com/scipy/scipy/blob/0c4cfab8b27d85785d28ec0e4b8d1ce536276c0e/scipy/interpolate/tests/test_rgi.py#L302-L314

In this test, `MyValue` is defined like this

```python
class MyValue:
    """
    Minimal indexable object
    """

    def __init__(self, shape):
        self.ndim = 2
        self.shape = shape
        self._v = np.arange(np.prod(shape)).reshape(shape)

    def __getitem__(self, idx):
        return self._v[idx]

    def __array_interface__(self):
        return None

    def __array__(self, dtype=None, copy=None):
        raise RuntimeError("No array representation")
```
        
Changing `RegularGridInterpolator` to have direct call to `array_namespace(*points, values)` results in the error

```
                # list, tuple, or arbitrary object
                try:
>                   array = np.asanyarray(array)
                            ^^^^^^^^^^^^^^^^^^^^
E                   ValueError: Invalid __array_interface__ value, must be a dict
```

inside of `array_namespace`. 

On `main`, this error sidestepped inside of `RegularGridInterpolator` here

https://github.com/scipy/scipy/blob/0c4cfab8b27d85785d28ec0e4b8d1ce536276c0e/scipy/interpolate/_rgi.py#L325-L334

we see that if `values` has an `ndim` `attr` then it never has `np.asarray` called on it, avoiding the failure due to invalid `__array_interface__`. 

I wasn't around to know why things are the way they are, but unfortunately this doesn't seem to be something that was accidentally supported that could be removed in the spirit of #18286. This appears to be some deliberately supported quirky behavior that presumably some users are relying on if it made it's way into the tests like this.

steppi replied:
Yes, this looks good. I realized laying in bed last night that my suggestion wasn't really satisfactory. I thought of something similar to what you ended up doing, but slightly more complicated because it only allowed duck dtypes `values` when `points` are all NumPy arrays. I don't think the extra complication is necessary, and prefer how you did it here.

steppi reviewed (commented):
Besides concerns about the bare `Exception` being caught. I think this looks good @ev-br.

ev-br committed:
ENH: interpolate/RGI: more careful handling of "duck-type values"

Distinguish between:

- *points have inconsistent namespaces -- then raise
- points are consistent, but values is not -- then raise
- points are consistent and the values is "duck-typed" -- then proceed

lucascolley commented on scipy/interpolate/_rgi.py:
```diff
@@ -281,17 +290,34 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+            xp = array_namespace(*points, values)
+        except Exception as e:
+            # either "duck-type" values or a user error?
+            xp = array_namespace(*points)
```
```suggestion
            xp = array_namespace(*points) # still forbid mixed namespaces in `points`
```

ev-br replied:
Thanks, this way it's _obviously better_ (tm). Taken along.

lucascolley reviewed (commented):
this handling looks good to me now with one extra comment, thanks Evgeni!

ev-br committed:
MAINT: interpolate: add a comment to RGI "duck-typed" dance

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

steppi reviewed (approved):
Looks good. Thanks Evgeni!
